raft.c

```diff
@@ -2575,26 +2488,54 @@ static void handleConfig(RedisRaftCtx *rr, RaftReq *req)
     RaftReqFree(req);
 }
 
-static RaftReqHandler RaftReqHandlers[] = {
-    NULL,
-    handleClusterInit,         /* RR_CLUSTER_INIT */
-    handleClusterJoin,         /* RR_CLUSTER_JOIN */
-    handleCfgChange,           /* RR_CFGCHANGE_ADDNODE */
-    handleCfgChange,           /* RR_CFGCHANGE_REMOVENODE */
-    handleAppendEntries,       /* RR_APPENDENTRIES */
-    handleRequestVote,         /* RR_REQUESTVOTE */
-    handleRedisCommand,        /* RR_REDISCOMMAND */
-    handleInfo,                /* RR_INFO */
-    handleSnapshot,            /* RR_SNAPSHOT */
-    handleDebug,              /* RR_DEBUG */
-    handleClientDisconnect,   /* RR_CLIENT_DISCONNECT */
-    handleShardGroupAdd,      /* RR_SHARDGROUP_ADD */
-    handleShardGroupsReplace, /* RR_SHARDGROUP_REPLACE */
-    handleShardGroupGet,      /* RR_SHARDGROUP_GET */
-    handleShardGroupLink,     /* RR_SHARDGROUP_LINK */
-    handleNodeShutdown,       /* RR_NODE_SHUTDOWN */
-    handleTransferLeader,     /* RR_TRANSFER_LEADER */
-    handleTimeoutNow,         /* RR_TIMEOUT_NOW */
-    handleConfig,             /* RR_CONFIG */
-    NULL
-};
+void handleFsyncCompleted(void *result)
+{
+    FsyncThreadResult *rs = result;
+    RedisRaftCtx *rr = &redis_raft;
+
+    rr->log->fsync_count++;
+    rr->log->fsync_total += rs->time;
+
+    int e = raft_flush(rr->raft, rs->fsync_index);
+    if (e == RAFT_ERR_SHUTDOWN) {
+        shutdownAfterRemoval(rr);
+    }
+    RedisModule_Assert(e == 0);
+
+    RedisModule_Free(rs);
+}
+
+void handleBeforeSleep(RedisRaftCtx *rr)
+{
+    int e;
+    raft_index_t flushed = 0;
+
+    if (rr->state == REDIS_RAFT_UP) {
+        if (!raft_is_leader(rr->raft)) {
+            return;
+        }
+
+        if (!rr->log->fsync) {
+            raft_index_t next = raft_get_index_to_sync(rr->raft);
+            if (next > 0) {
+                fflush(rr->log->file);
+            }
+            flushed = next;
+            goto out;
+        }
+
+        raft_index_t next = raft_get_index_to_sync(rr->raft);
```

Comment:
Everything here up to fsyncThreadAddTask() is identical, consider branching later?

Reply:
Refactored that part